homeassistant/components/tado/binary_sensor.py

```diff
@@ -125,3 +168,107 @@ def _async_update_device_data(self):
+
+        self._unique_id = f"{zone_variable} {zone_id} {tado.home_id}"
+
+        self._state = None
+        self._state_attributes = None
+        self._tado_zone_data = None
+
+    async def async_added_to_hass(self):
+        """Register for sensor updates."""
+
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                SIGNAL_TADO_UPDATE_RECEIVED.format(
+                    self._tado.home_id, "zone", self.zone_id
+                ),
+                self._async_update_callback,
+            )
+        )
+        self._async_update_zone_data()
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self.zone_name} {self.zone_variable}"
+
+    @property
+    def is_on(self):
+        """Return true if sensor is on."""
+        return self._state
+
+    @property
+    def device_class(self):
+        """Return the class of this sensor."""
+        if self.zone_variable == "early start":
+            return DEVICE_CLASS_POWER
+        if self.zone_variable == "link":
+            return DEVICE_CLASS_CONNECTIVITY
+        if self.zone_variable == "open window":
+            return DEVICE_CLASS_WINDOW
+        if self.zone_variable == "overlay":
+            return DEVICE_CLASS_POWER
+        if self.zone_variable == "power":
+            return DEVICE_CLASS_POWER
+        return None
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._state_attributes
+
+    @callback
+    def _async_update_callback(self):
+        """Update and write state."""
+        self._async_update_zone_data()
+        self.async_write_ha_state()
+
+    @callback
+    def _async_update_zone_data(self):
+        """Handle update callbacks."""
+        try:
+            self._tado_zone_data = self._tado.data["zone"][self.zone_id]
+        except KeyError:
+            return
+
+        if self.zone_variable == "power":
+            self._state = self._tado_zone_data.power
+
+        elif self.zone_variable == "link":
+            self._state = self._tado_zone_data.link
+
+        elif self.zone_variable == "overlay":
+            self._state = self._tado_zone_data.overlay_active
+            self._state_attributes = (
+                {"termination": self._tado_zone_data.overlay_termination_type}
+                if self._tado_zone_data.overlay_active
+                else {}
+            )
+
+        elif self.zone_variable == "early start":
+            self._state = self._tado_zone_data.preparation
+
+        elif self.zone_variable == "open window":
+            self._state = bool(
+                self._tado_zone_data.open_window
+                or self._tado_zone_data.open_window_detected
+            )
+            self._state_attributes = self._tado_zone_data.open_window_attr
```

Comment:
Please set all state attributes explicitly. We want to see the keys and their formatting.

Reply:
@MartinHjelmare excuse but I'm not following. Can you explain this? (if possible give an example or documentation link please)

Reply:
We want to see what attributes we're setting as state attributes. Please name the keys explicitly in the code. Example:
```py
ATTRIBUTE_MAP = {
    TADO_OPEN_WINDOW_ATTR: "open_window",
    ...
}

...

    for tado_attr, state_attr in ATTRIBUTE_MAP.items():
        self._state_attributes[state_attr] = self._tado_zone_data.open_window_attr[tado_attr]
```

Reply:
I've removed _state_attributes from binary sensors (these attributes aren't really useful).

Reply:
Please also mention this in the breaking change paragraph.

Reply:
Done:

> ## Breaking change
> Tado zone sensors will be replaced by their corresponding binary_sensors. Therefore, users will have to update lovelace panels, scripts and automations referring to the following sensors by their corresponding binary sensor (replace **sensor** with **binary_sensor**)
> 
> * sensor.%TADO_ZONE%_early_start will be replaced by binary_sensor.%TADO_ZONE%_early_start
> * sensor.%TADO_ZONE%_link will be replaced by binary_sensor.%TADO_ZONE%_link
> * sensor.%TADO_ZONE%_open_window will be replaced by binary_sensor.%TADO_ZONE%_open_window
> * sensor.%TADO_ZONE%_overlay will be replaced by binary_sensor.%TADO_ZONE%_overlay
> * sensor.%TADO_ZONE%_power will be replaced by binary_sensor.%TADO_ZONE%_power
> 
> The following state attributes have been removed:
> 
> * sensor.%TADO_ZONE%_overlay: termination
> * sensor.%TADO_ZONE%_open_window: open_window_attr